src/Transports.AspNetCore/WebSockets/BaseSubscriptionServer.Observer.cs

```diff
@@ -0,0 +1,80 @@
+namespace GraphQL.Server.Transports.AspNetCore.WebSockets;
+
+public abstract partial class BaseSubscriptionServer
+{
+    /// <summary>
+    /// Handles messages from the event source.
+    /// </summary>
+    private class Observer : IObserver<ExecutionResult>
+    {
+        private readonly BaseSubscriptionServer _server;
+        private readonly string _id;
+        private readonly bool _closeAfterOnError;
+        private readonly bool _closeAfterAnyError;
+        private int _done;
+
+        public Observer(BaseSubscriptionServer server, string id, bool closeAfterOnError, bool closeAfterAnyError)
+        {
+            _server = server;
+            _id = id;
+            _closeAfterOnError = closeAfterOnError;
+            _closeAfterAnyError = closeAfterAnyError;
+        }
+
+        public void OnCompleted()
+        {
+            if (Interlocked.Exchange(ref _done, 1) == 1)
+                return;
+            try
+            {
+                _ = _server.SendCompletedAsync(_id);
+            }
+            catch { }
+        }
+
+        public async void OnError(Exception error)
+        {
+            if (Thread.VolatileRead(ref _done) == 1)
+                return;
+            if (_closeAfterOnError && Interlocked.Exchange(ref _done, 1) == 1)
+                return;
+            try
+            {
+                if (error != null)
```

Comment:
Why does this null-check here? And how does it correlate with `if (_closeAfterOnError && Interlocked.Exchange(ref _done, 1) == 1)` condition ?

Reply:
We do not have control of whether a event source calls `OnError` with `null` or an actual exception.  They certainly should not pass in `null`.  Nevertheless, I wrote code to prevent a NRE in that case.  There is no requirement that a message be sent in this case, so it is valid from a protocol point of view.

`_closeAfterOnError` behaves just as it should, closing the connection if it is `true`.  There is no requirement that an error packet/message be sent first.

Reply:
Please add a comment about that to clarify intent.

Reply:
```suggestion
                // although error should never be null, if the event source does call OnError(null),
                // we will skip sending an error packet/message (allowed by spec)
                if (error != null)
```

Reply:
👍